update and detail MV and DT support (#3779)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/snowflake-configs.md

</File>

</VersionBlock>

<VersionBlock firstVersion="1.6">

## Dynamic Tables

[Dynamic Tables](https://docs.snowflake.com/en/user-guide/dynamic-tables-about) are Snowflake's version of Materialized Views.

### Parameters

dbt-snowflake requires the following parameters:

> - `TARGET_LAG`

Reviewer: ```suggestion
- `target_lag`
```

> - `WAREHOUSE`

Reviewer: ```suggestion
- `snowflake_warehouse`
```

To learn more about each parameter and what values it can take, see 

> the Snowflake docs page: [`CREATE DYNAMIC TABLE: Parameters`](https://docs.snowflake.com/en/sql-reference/sql/create-dynamic-table)

Reviewer: ```suggestion
the Snowflake docs page: [`CREATE DYNAMIC TABLE: Parameters`](https://docs.snowflake.com/en/sql-reference/sql/create-dynamic-table). Note that `downstream` for `target_lag` is not supported. 
```

### Usage

You can create a dynamic table by editing _one_ of these files:

- the SQL file for your model
- the `dbt_project.yml` configuration file

The following examples create a dynamic table:

<File name='models/YOUR_MODEL_NAME.sql'>

```sql
{{
  config(
    materialized = 'dynamic_table',

>     warehouse = 'MY_WAREHOUSE',

Reviewer: ```suggestion
    warehouse = 'snowflake_warehouse',
```

    target_lag = '10 minutes',
  )
}}
```

</File>

<File name='dbt_project.yml'>

```yaml
models:
  path:
    materialized: dynamic_table

>     warehouse: MY_WAREHOUSE

Reviewer: ```suggestion
    warehouse: snowflake_warehouse
```

</File>

</VersionBlock>

<VersionBlock firstVersion="1.6">

## Dynamic Tables

[Dynamic Tables](https://docs.snowflake.com/en/user-guide/dynamic-tables-about) are Snowflake's flavor of Materialized Views. The `CREATE DYNAMIC TABLE` ([docs](https://docs.snowflake.com/en/sql-reference/sql/create-dynamic-table)) statement requires the following parameters `TARGET_LAG` and `WAREHOUSE`, so a dbt-snowflake user must also provide these.

You can create a dynamic table by editing _one_ of these files:

- the SQL file for your model
- the `dbt_project.yml` configuration file

The following examples create a dynamic table:

<File name='models/YOUR_MODEL_NAME.sql'>

```sql
{{
  config(
    materialized = 'dynamic_table',
    warehouse = 'MY_WAREHOUSE',

>     target_lag = '10 min',

Reviewer 2: `10 minutes`. Supported durations are `second`, `minute`, `hour`, `day` and their plurals.

Reviewer 3: Note, "downsteam" is also also supported for target_lag: https://docs.snowflake.com/en/user-guide/dynamic-tables-refresh#understanding-dynamic-table-lag

Reviewer 2: Correct, `downstream` is now supported for Snowflake; however, that has not yet been added to `dbt-snowflake`.

Author: ```suggestion
    target_lag = '10 minutes',
```

  )
}}
```

</File>

<File name='dbt_project.yml'>

```yaml
models:
  path:
    materialized: dynamic_table
    warehouse: MY_WAREHOUSE

>     target_lag: '10 min'

Reviewer 2: `minutes`

Author: ```suggestion
    target_lag: '10 minutes'
```

</File>

</VersionBlock>

<VersionBlock firstVersion="1.6">

## Dynamic Tables

> 

Reviewer 2: There's no mention of `on_configuration_change`

> [Dynamic Tables](https://docs.snowflake.com/en/user-guide/dynamic-tables-about) are Snowflake's flavor of Materialized Views.

Reviewer: ```suggestion
[Dynamic Tables](https://docs.snowflake.com/en/user-guide/dynamic-tables-about) are Snowflake's version of Materialized Views.
```

### Parameters

dbt-snowflake requires the following parameters:

> 

Reviewer 2: The parameters are lowercase and the second one is incorrect:
```
- target_lag
- snowflake_warehouse
```

- `TARGET_LAG`
- `WAREHOUSE`

> To learn more about each parameter and what values it can take, see 

Reviewer 2: We don't support `downstream` for `target_lag` yet. This will appear in the linked docs, so it should be called out here that we don't support it.

> the Snowflake docs page: [`CREATE DYNAMIC TABLE: Parameters`](https://docs.snowflake.com/en/sql-reference/sql/create-dynamic-table)

Author: @sfc-gh-slionheart @mikealfare 
how's this for calling out the parameters? I'd rather not have to duplicate y'all's great docs page, but could buy the argument that this doesn't result in the best end-user experience

Reviewer 2: I think I'm fine with that, though I thought the parameter for warehouse was snowflake_warehouse ([here](https://docs.getdbt.com/reference/resource-configs/snowflake-configs#configuring-virtual-warehouses)). Also, the arguments are lowercase, so `target_lag` and `snowflake_warehouse`. They come in as strings that are case-sensitive in python and we don't adjust the case.

### Usage

You can create a dynamic table by editing _one_ of these files:

- the SQL file for your model
- the `dbt_project.yml` configuration file

The following examples create a dynamic table:

<File name='models/YOUR_MODEL_NAME.sql'>

```sql
{{
  config(
    materialized = 'dynamic_table',

>     warehouse = 'MY_WAREHOUSE',

Reviewer 2: snowflake_warehouse

    target_lag = '10 minutes',
  )
}}
```

</File>

<File name='dbt_project.yml'>

```yaml
models:
  path:
    materialized: dynamic_table

>     warehouse: MY_WAREHOUSE

Reviewer 2: snowflake_warehouse

    target_lag: '10 minutes'
```

</File>

### Limitations

#### Changing materialization to and from "dynamic_table"

> Swapping an already materialized model to be a dynamic table and vise versa. The workaround is the manually drop the existing materialization in the data warehouse before calling `dbt run` again.

Reviewer: ```suggestion
Swapping an already materialized model to be a dynamic table and vice versa. The workaround is manually dropping the existing materialization in the data warehouse before calling `dbt run` again.
```

> To illustrate, assume for the example model below, `my_model`, that it has already been materialized to the underlying data platform via `dbt run`. If a user then changes the model's config to be `materialized="dynamic_table"`, they will get an error. The workaround is to execute `DROP TABLE my_model` on the data warehouse before trying the model again.

Reviewer: ```suggestion
For example, assume for the example model below, `my_model`, has already been materialized to the underlying data platform via `dbt run`. If a user then changes the model's config to be `materialized="dynamic_table"`, they will get an error. The workaround is to execute `DROP TABLE my_model` on the data warehouse before trying the model again.
```

<File name='my_model.sql'>

```yaml

{{ config(
    materialized="table" # or any model type eg view, incremental
) }}

```

</File>

> #### Altering `target_lag` or `warehouse`

Reviewer: ```suggestion
```

> Currently, changing either the `target_lag` or `warehouse` parameters for a pre-existing Dynamic Table (DT) will result in the Dynamic Table being dropped and replaced, rather than altered.

Reviewer: ```suggestion
```

> Currently, changing either the `target_lag` or `warehouse` parameters for a pre-existing Dynamic Table (DT) will result in the Dynamic Table being dropped and replaced, rather than altered.

Reviewer 2: We got this into 1.6, so this is no longer a limitation.
